### DISCUSSION TITLE
controller-activation-and-dependency-injection-in-asp-net-core-mvc/

### DISCUSSION THREAD

**Discussion by andrewlock** [andrewlock/blog-comments#217] [Announcements]
# controller-activation-and-dependency-injection-in-asp-net-core-mvc/

In this post, I'll describe how controllers are created in ASP.NET Core MVC using the IControllerActivator and the options available out of the box.

https://andrewlock.net/controller-activation-and-dependency-injection-in-asp-net-core-mvc/

**Comment by OsamaAlRashed** [comment]
Thanks for this amazing article!
but why the controllers are not created by the DI container by default, what the benefits of "IControllerActivitor"?

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at February 24 2018, 09:23, in reply to [Elon Must](https://disqus.com/by/drsensor/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/217#discussioncomment-6042687)</em>

---

<p>Can I suggest you checkout StructureMap or Scrutor - these both already do what you want :)<br>Structor in particular is probably very close to what you're after:<br><a href="https://github.com/khellang/Scrutor" rel="nofollow noopener" title="https://github.com/khellang/Scrutor">https://github.com/khellang/Scrutor</a></p>

[a comment by an automated account is not shown]

**Comment by zhang1446802857** [comment]
I am attempting to use attribute injection for the controller. If using AddControllersAsServices, the effect can still be achieved without using IControllerActivator.
